File: src/http-transport.ts
/**
 * HTTP Transport for MVP client
 */

import type { CmdReq, CmdResp, ErrorInfo } from './envelope';
import { HEADER_PREFIX, newCmdResp } from './envelope';

/**
 * Transporter interface for sending commands
 */
export interface Transporter {
  /**
   * Transports a command and returns the response
   */
  transportCmd(
    cmdName: string,
    contentType: string,
    payload: unknown
  ): Promise<unknown>;
}

/**
 * EnvelopeTransporter interface for sending commands with headers
 */
export interface EnvelopeTransporter {
  /**
   * Transports a command request with headers and returns a response with headers
   */
  transportCmdReq(req: CmdReq, contentType: string): Promise<CmdResp>;
}

/**
 * Encoder interface for serializing/deserializing data
 */
export interface Encoder {
  /** MIME type for the encoding */
  mimeType: string;
  /** Encodes data to string or bytes */
  encode(data: unknown): string | Uint8Array;
  /** Decodes string or bytes to data */
  decode<T>(data: string | Uint8Array): T;
}

/**
 * JSON encoder implementation
 */
export const jsonEncoder: Encoder = {
  mimeType: 'application/json',
  encode: (data: unknown) => JSON.stringify(data),
  decode: <T>(data: string | Uint8Array): T => {
    const str = typeof data === 'string' ? data : new TextDecoder().decode(data);
    return JSON.parse(str) as T;
  },
};

/**
 * FetchOptions for customizing fetch behavior
 */
export interface FetchOptions {
  /** Custom fetch implementation (useful for Node.js or testing) */
  fetch?: typeof fetch;
  /** Additional headers to include in all requests */
  headers?: Record<string, string>;
  /** Request credentials mode */
  credentials?: RequestCredentials;
  /** Request mode */
  mode?: RequestMode;
}

/**
 * HttpTransporter implements HTTP-based command transport
 */
export class HttpTransporter implements Transporter, EnvelopeTransporter {
  private url: string;
  private path: string;
  private fetchFn: typeof fetch;
  private defaultHeaders: Record<string, string>;
  private credentials?: RequestCredentials;
  private mode?: RequestMode;

  constructor(
    url: string,
    pkgPath: string,
    options: FetchOptions = {}
  ) {
    if (!url) {
      throw new Error('missing url');
    }
    if (!url.startsWith('http://') && !url.startsWith('https://')) {
      throw new Error('invalid url: must start with http:// or https://');
    }
    if (!pkgPath) {
      throw new Error('missing package path');
    }

    this.url = url;
    this.path = url + pkgPath;
    this.fetchFn = options.fetch || globalThis.fetch;
    this.defaultHeaders = options.headers || {};
    this.credentials = options.credentials;
    this.mode = options.mode;

    if (!this.fetchFn) {
      throw new Error('fetch is not available. Please provide a fetch implementation.');
    }
  }

  /**
   * Gets the full URL for requests
   */
  getUrl(): string {
    return this.path;
  }

  /**
   * Transports a raw command without envelope
   */
  async transportCmd(
    cmdName: string,
    contentType: string,
    payload: unknown
  ): Promise<unknown> {
    if (!cmdName) {
      throw new Error('missing command name');
    }
    if (!contentType) {
      throw new Error('missing content type');
    }

    const body = typeof payload === 'string' ? payload : JSON.stringify(payload);

    const response = await this.fetchFn(this.path, {
      method: 'POST',
      headers: {
        ...this.defaultHeaders,
        'Content-Type': contentType,
        'x-mainvec-cmd': cmdName,
      },
      body,
      credentials: this.credentials,
      mode: this.mode,
    });

    if (!response.ok) {
      const errorMsg = response.headers.get('x-mainvec-error');
      if (errorMsg) {
        throw new Error(`error: ${errorMsg}`);
      }
      throw new Error(`unexpected status code: ${response.status}`);
    }

    const responseText = await response.text();
    try {
      return JSON.parse(responseText);
    } catch {
      return responseText;
    }
  }

  /**
   * Transports a command request with headers and returns a response with headers
   */
  async transportCmdReq(req: CmdReq, contentType: string): Promise<CmdResp> {
    if (!req) {
      throw new Error('missing command request');
    }
    if (!req.cmd) {
      throw new Error('missing command name');
    }
    if (!contentType) {
      throw new Error('missing content type');
    }

    // Build headers
    const headers: Record<string, string> = {
      ...this.defaultHeaders,
      'Content-Type': contentType,
      'x-mainvec-cmd': req.cmd,
    };

    // Add custom headers with prefix
    for (const [key, value] of Object.entries(req.headers)) {
      headers[HEADER_PREFIX + key] = value;
    }

    // Encode payload
    const body = typeof req.payload === 'string' 
      ? req.payload 
      : JSON.stringify(req.payload);

    const response = await this.fetchFn(this.path, {
      method: 'POST',
      headers,
      body,
      credentials: this.credentials,
      mode: this.mode,
    });

    // Read response body
    const responseText = await response.text();
    let payload: unknown;
    try {
      payload = JSON.parse(responseText);
    } catch {
      payload = responseText;
    }

    // Build CmdResp
    const cmdResp = newCmdResp(payload);

    // Extract response headers with prefix
    response.headers.forEach((value, key) => {
      const lowerKey = key.toLowerCase();
      if (lowerKey.startsWith(HEADER_PREFIX)) {
        const headerKey = lowerKey.slice(HEADER_PREFIX.length);
        cmdResp.headers[headerKey] = value;
      }
    });

    // Check for errors
    if (!response.ok) {
      const errorMsg = response.headers.get('x-mainvec-error') || responseText;
      const error: ErrorInfo = {
        code: `http_${response.status}`,
        message: errorMsg,
      };
      cmdResp.error = error;
    }

    return cmdResp;
  }
}

/**
 * Creates a new HttpTransporter
 */
export function newHttpTransporter(
  url: string,
  pkgPath: string,
  options?: FetchOptions
): HttpTransporter {
  return new HttpTransporter(url, pkgPath, options);
}

File: mvp/http_transport.go
package mvp

import (
	"context"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"strings"

	//_ "github.com/mainvec/mvp/mvpgo/mvp/util/protojson"
	_ "github.com/mainvec/ugo/oencoding/json"
)

var _ Transporter = &HttpTransporter{}
var _ EnvelopeTransporter = &HttpTransporter{}

type HttpTransporter struct {
	url    string
	path   string
	client *http.Client
}

func NewHttpTransporter(url string, pkgPath string) (*HttpTransporter, error) {
	return NewHttpTransporterWithClient(url, pkgPath, &http.Client{})
}

func NewHttpTransporterWithClient(url string, pkgPath string, httpClient *http.Client) (*HttpTransporter, error) {
	if len(url) == 0 {
		return nil, fmt.Errorf("missing url")
	}
	if !strings.HasPrefix(url, "http://") && !strings.HasPrefix(url, "https://") {
		return nil, fmt.Errorf("invalid url")
	}
	if len(pkgPath) == 0 {
		return nil, fmt.Errorf("missing package path")
	}
	if httpClient == nil {
		return nil, fmt.Errorf("missing http client")
	}

	return &HttpTransporter{
		url:    url,
		path:   url + pkgPath,
		client: httpClient,
	}, nil
}

func (t *HttpTransporter) TransportCmd(ctx context.Context, cmdName string, contentType string, cmdData io.ReadCloser) (io.ReadCloser, error) {
	if len(cmdName) == 0 {
		return nil, fmt.Errorf("missing command name")
	}
	if len(contentType) == 0 {
		return nil, fmt.Errorf("missing content type")
	}
	if cmdData == nil {
		return nil, fmt.Errorf("missing command data")
	}

	req, err := http.NewRequest("POST", t.path, cmdData)
	if err != nil {
		return nil, fmt.Errorf("failed to create http request: %w", err)
	}
	req.Header.Set("Content-Type", contentType)
	req.Header.Set("x-mainvec-cmd", cmdName)
	resp, err := t.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send http request: %w", err)
	}
	if resp.StatusCode != http.StatusOK {
		if resp.Header.Get("x-mainvec-error") != "" {
			return nil, fmt.Errorf("error: %s", resp.Header.Get("x-mainvec-error"))
		}
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}
	return resp.Body, nil
}

// TransportCmdReq sends a command request with headers and returns a response with headers
func (t *HttpTransporter) TransportCmdReq(ctx context.Context, cmdReq *CmdReq, contentType string) (*CmdResp, error) {
	if cmdReq == nil {
		return nil, fmt.Errorf("missing command request")
	}
	if len(cmdReq.Cmd) == 0 {
		return nil, fmt.Errorf("missing command name")
	}
	if len(contentType) == 0 {
		return nil, fmt.Errorf("missing content type")
	}

	req, err := http.NewRequestWithContext(ctx, "POST", t.path, io.NopCloser(strings.NewReader(string(cmdReq.Payload))))
	if err != nil {
		return nil, fmt.Errorf("failed to create http request: %w", err)
	}

	// Set standard headers
	req.Header.Set("Content-Type", contentType)
	req.Header.Set("x-mainvec-cmd", cmdReq.Cmd)

	// Set custom headers with prefix
	for k, v := range cmdReq.Headers {
		req.Header.Set(HeaderPrefix+k, v)
	}

	resp, err := t.client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("failed to send http request: %w", err)
	}
	defer resp.Body.Close()

	// Read response body
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read response body: %w", err)
	}

	// Build CmdResp
	cmdResp := &CmdResp{
		Headers: make(map[string]string),
		Payload: respBody,
	}

	// Extract response headers with prefix
	for k, v := range resp.Header {
		lowerKey := strings.ToLower(k)
		if strings.HasPrefix(lowerKey, HeaderPrefix) && len(v) > 0 {
			headerKey := strings.TrimPrefix(lowerKey, HeaderPrefix)
			cmdResp.Headers[headerKey] = v[0]
		}
	}

	// Check for errors
	if resp.StatusCode != http.StatusOK {
		errorMsg := resp.Header.Get("x-mainvec-error")
		if errorMsg == "" {
			errorMsg = string(respBody)
		}
		cmdResp.Error = &ErrorInfo{
			Code:    fmt.Sprintf("http_%d", resp.StatusCode),
			Message: errorMsg,
		}
	}

	return cmdResp, nil
}

func (h *PackageHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	cmdName := r.Header.Get("x-mainvec-cmd")
	slog.Debug("received cmd", "x-mainvec-cmd", cmdName)
	if cmdName == "" {
		http.Error(w, "missing x-mainvec-cmd header", http.StatusBadRequest)
		return
	}

	encoder := r.Header.Get("Content-Type")
	slog.Debug("received cmd", "Content-Type", encoder)
	if encoder == "" {
		http.Error(w, "missing Content-Type header", http.StatusBadRequest)
		return
	}

	// Extract request headers with prefix
	reqHeaders := make(map[string]string)
	for k, v := range r.Header {
		lowerKey := strings.ToLower(k)
		if strings.HasPrefix(lowerKey, HeaderPrefix) && len(v) > 0 {
			headerKey := strings.TrimPrefix(lowerKey, HeaderPrefix)
			reqHeaders[headerKey] = v[0]
		}
	}

	// Read request body
	defer r.Body.Close()
	payload, err := io.ReadAll(r.Body)
	if err != nil {
		slog.Error("failed to read request body", "error", err.Error())
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}

	// Build CmdReq and call ServeCmdReq
	cmdReq := &CmdReq{
		Cmd:     cmdName,
		Headers: reqHeaders,
		Payload: payload,
	}

	cmdResp := h.ServeCmdReq(r.Context(), cmdReq, encoder)

	// Set response headers with prefix
	for k, v := range cmdResp.Headers {
		w.Header().Set(HeaderPrefix+k, v)
	}

	// Handle error response
	if cmdResp.HasError() {
		slog.Error("command failed", "code", cmdResp.Error.Code, "message", cmdResp.Error.Message)
		w.Header().Set("x-mainvec-error", cmdResp.Error.Message)
		w.Header().Set("Content-Type", encoder)
		http.Error(w, cmdResp.Error.Message, http.StatusBadRequest)
		return
	}

	// Write success response
	w.Header().Set("Content-Type", encoder)
	w.WriteHeader(http.StatusOK)
	_, err = w.Write(cmdResp.Payload)
	if err != nil {
		slog.Error("failed to write response", "error", err.Error())
	}
}
